src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py

```diff
@@ -158,6 +158,19 @@ def is_large_local_data(self) -> bool:
 
         return len(self.data) * len(self.output) >= ARRAY_BIND_THRESHOLD
 
+    @property
+    def is_contain_illegal_null_value(self) -> bool:
+        from snowflake.snowpark._internal.analyzer.analyzer import ARRAY_BIND_THRESHOLD
+
+        rows_to_compare = min(
+            ARRAY_BIND_THRESHOLD // len(self.output) + 1, len(self.data)
+        )
+        for i in range(rows_to_compare):
+            for j in range(len(self.output)):
+                if self.data[i][j] is None and not self.output[j].nullable:
+                    return True
```

Comment:
Potentially a lot fewer comparisons if you restructured like this.
```suggestion
        for j in range(len(self.output)):
            if not self.output[j].nullable:
                for i in range(rows_to_compare):
                    if self.data[i][j] is None:
                        return True
```

Reply:
you are right, will change